src/qutip_cupy/dense.py

```diff
@@ -0,0 +1,135 @@
+"""
+This module contains the ``CuPyDense`` class and associated function
+conversion and specializations for registration with QuTiP's data layer.
+"""
+
+import numbers
+
+import cupy as cp
+from qutip.core import data
+
+
+class CuPyDense(data.Data):
+    """
+        This class provides a dense matrix backend for QuTiP.
+        Matrices are stored internally in a CuPy array on a GPU.
+        If you have many GPUs you can set GPU ``i``
+        by calling ``cp.cuda.Device(i).use()`` before construction.
+
+        Parameters
+        ----------
+        data: array-like
+            Data to be stored.
+        shape: (int, int)
+            Defaults to ``None``. If ``None`` will infer the shape from ``data``,
+            else it will set the shape for the internal CuPy array.
+        copy: bool
+            Defaults to ``True``. Whether to make a copy of
+            the elements in ``data`` or not.
+        dtype:
+            Data type specifier. Either ``cp.complex128`` or ``cp.complex64``
+
```

Comment:
```suggestion
    This class provides a dense matrix backend for QuTiP.
    Matrices are stored internally in a CuPy array on a GPU.
    If you have many GPUs you can set GPU ``i``
    by calling ``cp.cuda.Device(i).use()`` before construction.

    Parameters
    ----------
    data: array-like
        Data to be stored.
    shape: (int, int)
        Defaults to ``None``. If ``None`` will infer the shape from ``data``,
        else it will set the shape for the internal CuPy array.
    copy: bool
        Defaults to ``True``. Whether to make a copy of
        the elements in ``data`` or not.
    dtype:
        Data type specifier. Either ``cp.complex128`` or ``cp.complex64``
```
Properly indent docstring.